File: TextosJapones/src/main/java/org/jisho/textosJapones/database/mysql/Backup.java
package org.jisho.textosJapones.database.mysql;

import javafx.stage.DirectoryChooser;
import javafx.stage.FileChooser;
import org.jisho.textosJapones.components.notification.AlertasPopup;
import org.jisho.textosJapones.components.notification.Notificacoes;
import org.jisho.textosJapones.controller.MenuPrincipalController;
import org.jisho.textosJapones.model.enums.Notificacao;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;

public class Backup {

    private static final Logger LOGGER = LoggerFactory.getLogger(Backup.class);

    private static final String SEPARATOR = File.separator;
    private static String MYSQL_PATH;

    private static void criaPasta(String caminho) {
        File arquivo = new File(caminho);
        if (!arquivo.exists())
            arquivo.mkdir();
    }

    private static void carregaCampos() throws Exception {
        ConexaoMysql.getDadosConexao();
        if (ConexaoMysql.getDataBase().isEmpty())
            throw new Exception("Erro, não existe nenhuma base nos arquivos de configuração.");

        if (ConexaoMysql.getCaminhoMysql().isEmpty())
            throw new Exception("Erro, caminho do mysql não definido no arquivo de configuração.");

        File arquivo = new File(ConexaoMysql.getCaminhoMysql());
        if (!arquivo.exists())
            throw new Exception("Erro, caminho do mysql não encontrado.");

        MYSQL_PATH = ConexaoMysql.getCaminhoMysql() + SEPARATOR + "bin";
    }

    private static String processaExportacao(File caminho) throws IOException {
        criaPasta(caminho + SEPARATOR);
        String commando = MYSQL_PATH + SEPARATOR + "mysqldump.exe";
        long time1, time2, time;
        time1 = System.currentTimeMillis();
        ProcessBuilder pb = new ProcessBuilder(commando, "--user=" + ConexaoMysql.getUser(),
                "--password=" + ConexaoMysql.getPassword(), ConexaoMysql.getDataBase(),
                "--result-file=" + caminho.getPath() + SEPARATOR + ConexaoMysql.getDataBase() + ".sql");

        pb.start();
        time2 = System.currentTimeMillis();
        time = (time2 - time1) / 1000;

        return "Backup realizado com sucesso. Tempo decorrido " + time + "s.\n" + caminho.getPath() + SEPARATOR
                + "Backup" + SEPARATOR + ConexaoMysql.getDataBase() + ".sql";
    }

    private static String processaImportacao(File arquivo) throws IOException {
        String commando = MYSQL_PATH + SEPARATOR + "mysql.exe";
        long time1, time2, time;
        time1 = System.currentTimeMillis();

        ProcessBuilder pb = new ProcessBuilder(commando, "--user=" + ConexaoMysql.getUser(),
                "--password=" + ConexaoMysql.getPassword(), ConexaoMysql.getDataBase(), "-e",
                " source " + arquivo.getAbsolutePath());

        pb.start();

        time2 = System.currentTimeMillis();
        time = (time2 - time1) / 1000;

        return "Backup importado com sucesso. Tempo decorrido " + time + "s.";
    }

    public static void importarBackup(MenuPrincipalController cnt) {
        FileChooser backup = new FileChooser();
        backup.setInitialDirectory(new File(System.getProperty("user.home")));
        backup.setTitle("Carregar o backup.");
        backup.getExtensionFilters().add(new FileChooser.ExtensionFilter("Arquivo de texto (*.sql)", "*.sql"));
        File arquivo = backup.showOpenDialog(null);

        if (arquivo == null)
            cnt.cancelaBackup();
        else {
            try {
                carregaCampos();
                String resultado = processaImportacao(arquivo);
                cnt.importaConcluido(false);
                Notificacoes.notificacao(Notificacao.SUCESSO, "Backup realizado com sucesso", resultado);
            } catch (IOException e) {
                
                LOGGER.error(e.getMessage(), e);
                cnt.importaConcluido(true);
                AlertasPopup.ErroModal("Erro ao tentar gerar backup da base", e.getStackTrace().toString());
            } catch (Exception e) {
                
                LOGGER.error(e.getMessage(), e);
                cnt.importaConcluido(true);
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao realizar o backup",
                        "Arquivo de configuração não encontrado ou nome da base não configurada.");
            }
        }
    }

    public static void exportarBackup(MenuPrincipalController cnt) {
        DirectoryChooser backup = new DirectoryChooser();
        backup.setInitialDirectory(new File(System.getProperty("user.home")));
        backup.setTitle("Carregar o backup.");
        File arquivo = backup.showDialog(null);

        if (arquivo == null)
            cnt.cancelaBackup();
        else {
            try {
                carregaCampos();
                String resultado = processaExportacao(arquivo);
                cnt.exportaConcluido(false);
                Notificacoes.notificacao(Notificacao.SUCESSO, "Backup realizado com sucesso", resultado);
            } catch (IOException e) {
                
                LOGGER.error(e.getMessage(), e);
                cnt.exportaConcluido(true);
                AlertasPopup.ErroModal("Erro ao tentar gerar backup da base", e.getStackTrace().toString());
            } catch (Exception e) {
                
                LOGGER.error(e.getMessage(), e);
                cnt.exportaConcluido(true);
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao realizar o backup",
                        "Arquivo de configuração não encontrado ou nome da base não configurada.");
            }
        }
    }

}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/database/mysql/Backup.kt
package br.com.fenix.processatexto.database.mysql

import br.com.fenix.processatexto.components.notification.AlertasPopup
import br.com.fenix.processatexto.components.notification.Notificacoes
import br.com.fenix.processatexto.controller.MenuPrincipalController
import br.com.fenix.processatexto.model.enums.Notificacao
import br.com.fenix.processatexto.util.configuration.Configuracao
import javafx.stage.DirectoryChooser
import javafx.stage.FileChooser
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException


object Backup {
    private val LOGGER = LoggerFactory.getLogger(Backup::class.java)
    private val SEPARATOR = File.separator
    private var MYSQL_PATH: String? = null

    private fun criaPasta(caminho: String) {
        val arquivo = File(caminho)
        if (!arquivo.exists())
            arquivo.mkdir()
    }

    @Throws(Exception::class)
    private fun carregaCampos() {
        if (Configuracao.database.isEmpty())
            throw Exception("Erro, não existe nenhuma base nos arquivos de configuração.")

        if (Configuracao.caminhoMysql.isEmpty())
            throw Exception("Erro, caminho do mysql não definido no arquivo de configuração.")

        val arquivo = File(Configuracao.caminhoMysql)

        if (!arquivo.exists())
            throw Exception("Erro, caminho do mysql não encontrado.")

        MYSQL_PATH = (Configuracao.caminhoMysql + SEPARATOR) + "bin"
    }

    @Throws(IOException::class)
    private fun processaExportacao(caminho: File): String {
        criaPasta(caminho.path + SEPARATOR)
        val commando = MYSQL_PATH + SEPARATOR + "mysqldump.exe"
        val time: Long
        val time1: Long = System.currentTimeMillis()
        val pb = ProcessBuilder(
            commando, "--user=" + Configuracao.user,
            "--password=" + Configuracao.password, Configuracao.database,
            "--result-file=" + caminho.path + SEPARATOR + Configuracao.database + ".sql"
        )
        pb.start()
        val time2: Long = System.currentTimeMillis()
        time = (time2 - time1) / 1000
        return "Backup realizado com sucesso. Tempo decorrido ${time}s. ${caminho.path}${SEPARATOR}Backup$SEPARATOR${Configuracao.database}.sql"
    }

    @Throws(IOException::class)
    private fun processaImportacao(arquivo: File): String {
        val commando = MYSQL_PATH + SEPARATOR + "mysql.exe"
        val time: Long
        val time1: Long = System.currentTimeMillis()
        val pb = ProcessBuilder(
            commando, "--user=" + Configuracao.user,
            "--password=" + Configuracao.password, Configuracao.database, "-e",
            " source " + arquivo.absolutePath
        )
        pb.start()
        val time2: Long = System.currentTimeMillis()
        time = (time2 - time1) / 1000
        return "Backup importado com sucesso. Tempo decorrido " + time + "s."
    }

    fun importarBackup(cnt: MenuPrincipalController) {
        val backup = FileChooser()
        backup.initialDirectory = File(System.getProperty("user.home"))
        backup.title = "Carregar o backup."
        backup.extensionFilters.add(FileChooser.ExtensionFilter("Arquivo de texto (*.sql)", "*.sql"))
        val arquivo: File = backup.showOpenDialog(null)
        if (arquivo == null)
            cnt.cancelaBackup()
        else {
            try {
                carregaCampos()
                val resultado = processaImportacao(arquivo)
                cnt.importaConcluido(false)
                Notificacoes.notificacao(Notificacao.SUCESSO, "Backup realizado com sucesso", resultado)
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
                cnt.importaConcluido(true)
                AlertasPopup.ErroModal("Erro ao tentar gerar backup da base", e.stackTrace.toString())
            } catch (e: Exception) {
                LOGGER.error(e.message, e)
                cnt.importaConcluido(true)
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao realizar o backup", "Arquivo de configuração não encontrado ou nome da base não configurada.")
            }
        }
    }

    fun exportarBackup(cnt: MenuPrincipalController) {
        val backup = DirectoryChooser()
        backup.initialDirectory = File(System.getProperty("user.home"))
        backup.title = "Carregar o backup."
        val arquivo: File = backup.showDialog(null)
        if (arquivo == null)
            cnt.cancelaBackup()
        else {
            try {
                carregaCampos()
                val resultado = processaExportacao(arquivo)
                cnt.exportaConcluido(false)
                Notificacoes.notificacao(Notificacao.SUCESSO, "Backup realizado com sucesso", resultado)
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
                cnt.exportaConcluido(true)
                AlertasPopup.ErroModal("Erro ao tentar gerar backup da base", e.stackTrace.toString())
            } catch (e: Exception) {
                LOGGER.error(e.message, e)
                cnt.exportaConcluido(true)
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao realizar o backup", "Arquivo de configuração não encontrado ou nome da base não configurada.")
            }
        }
    }
}